src/java/org/apache/cassandra/service/ReplicaFilteringProtection.java

```diff
@@ -239,7 +253,25 @@ public Row onMergedRows(Row merged, Row[] versions)
                 {
                     // cache the row versions to be able to regenerate the original row iterator
                     for (int i = 0; i < versions.length; i++)
+                    {
                         builders[i].addRow(versions[i]);
+                        rowsCached++;
+
+                        if (rowsCached == cachedRowsFailThreshold + 1)
+                        {
+                            throw new TooManyCachedRowsException(cachedRowsFailThreshold, command);
+                        }
+                        else if (rowsCached == cachedRowsWarnThreshold + 1)
+                        {
+                            String message =
+                                String.format("Replica filtering protection materialized %d rows during query %s. " +
```

Comment:
self nit: I'll probably change this to say "...materialized more than %d rows...", using `cachedRowsWarnThreshold` instead of `rowsCached`. At least the former indicates the threshold clearly.